traits/observers/_list_item_observer.py

```diff
@@ -0,0 +1,221 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+from traits.observers._list_change_event import list_event_factory
+from traits.observers._i_observer import IObserver
+from traits.observers._trait_event_notifier import TraitEventNotifier
+from traits.observers._observe import add_or_remove_notifiers
+from traits.observers._observer_change_notifier import ObserverChangeNotifier
+from traits.trait_list_object import TraitList
+
+
+@IObserver.register
+class ListItemObserver:
+    """ Observer for observing mutations on a list.
+
+    Parameters
+    ----------
+    notify : boolean
+        Whether to notify for changes.
+    optional : boolean
+        If False, this observer will complain if the incoming object is not
+        an observable list. If True and the incoming object is not a list, this
+        observer will do nothing. Useful for the 'items' keyword in the text
+        parser, where the source container type is ambiguous.
+    """
+
+    def __init__(self, *, notify, optional):
+        self._notify = notify
+        self._optional = optional
+
+    def __hash__(self):
+        """ Return a hash of this object."""
+        return hash((type(self), self.notify, self.optional))
+
+    def __eq__(self, other):
+        """ Return true if this observer is equal to the given one."""
+        return (
+            type(self) is type(other)
+            and self.notify == other.notify
+            and self.optional == other.optional
+        )
+
+    @property
+    def notify(self):
```

Comment:
What's the rationale for putting these attributes behind properties? I'd suggest keeping it simple and leaving them as plain attributes.

Reply:
I was hoping make these attributes immutable, to be defensive (against human errors/oversights).

Reply:
I think that only makes sense in Python where there's a higher-than-usual risk of attributes being overwritten (e.g., perhaps because there was a previous API that did this). For this particular situation, I don't think this type of accident is any more likely than any of the other possible accidents that we're not protecting against, so I think this is unnecessary extra boilerplate.

Reply:
> a higher-than-usual risk of attributes being overwritten

Or perhaps also in cases where the consequences of accidental overwriting would be worse than usual.

Reply:
These attributes are special because they are used in `__hash__`, therefore I am making an exclusive protection against mutating them. I am happy to relax it, just wanted to clarify in case that may change your view, as I realized I have not included the mention of hashing in my previous comment.

Reply:
>  am happy to relax it, just wanted to clarify in case that may change your view, as I realized I have not included the mention of hashing in my previous comment.

Just to clarify that I am not proceeding with this change yet until I get a confirmation.

Reply:
Let's simplify the code. I don't think this is a failure mode that's likely enough to make it worth guarding against.

Reply:
OK. I don't mind the boilerplate and I think defensiveness is good, but I will remove it nonetheless.

Reply:
Yes, defensiveness is good, but this is one isolated piece of defensiveness that's at a level that's more low risk than many other pieces we're not addressing: it's like having a lock on one window in a house while you leave all the other windows wide open. If we applied this level of defensiveness everywhere there would be extra boilerplate everywhere, and much of the advantage we get from programming in a low boilerplate language would be gone.